### PR TITLE
Fix SectionNode claiming edge cases

### PR DESCRIPTION
**Pull request by APickledWalrus**
### Problem
There are some edge cases with SectionNode claiming related to Functions and EffectSections. Currently, functions can always be used with sections, even if no parameter claims the section. This results in random unparsed code. EffectSections also do not properly work with ExpressionSections. The "sectionNode" of the SectionContext is always null, meaning there is nothing to claim (even if one is present). This can result in unclaimed/unparsed code.

#### For Functions
I simply added a check similar to the one performed below for Statement that ensures the Section has been claimed. I tweaked the error message to better fit functions. It does not currently attempt to try parsing as an EffectSection or Statement. This is partly due to the logging setup, and I believe that syntax conflicting with function calls is not currently supported anyways.

#### For EffectSections
I added a new `EffectSection#parse` override that supports specifying whether the `sectionNode` parameter can be claimed by the parsed EffectSection. This uses an internal ParserInstance context for passing this information around.

### Testing Completed
A test for the Function edge case was added.

### Supporting Information

---
**Completes:** none 
**Related:** none

### CODE REVIEW

File: src/main/java/ch/njol/skript/lang/EffectSection.java [hunk 1]
```diff
@@ -52,15 +69,48 @@ public abstract boolean init(Expression<?>[] expressions,
...
+				return null;
+			}
+			return parsed;
 		});
+
+		effectSectionContext.isNodeForEffectSection = wasNodeForEffectSection;
+		return effectSection;
+	}
+
+	private static class EffectSectionContext extends ParserInstance.Data {
```
Comment (Absolutionism) [inline src/main/java/ch/njol/skript/lang/EffectSection.java:103]: Could this extend `SectionContext`? instead of needing to grab both this and `SectionContext`

Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/lang/EffectSection.java:103]: Unlikely as other places (outside of EffectSection) will be expecting SectionContext (and calling SectionContext.class). I could have included this field on SectionContext itself, but I preferred this way to keep it internal to EffectSection.

Comment (Absolutionism) [reply src/main/java/ch/njol/skript/lang/EffectSection.java:103]: Well I meant having `EffectSectionContext` extend `SectionContext`, so when doing `getData(EffectSectionContext.class)` you would have access to both.

Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/lang/EffectSection.java:103]: Right, but other areas of Skript would still try to access it under `getData(SectionContext.class)`, which would result in the expected values not being present.
